From fcc237f113c596abfc4e95ddbe67dde5782d13f1 Mon Sep 17 00:00:00 2001
From: Aleix Pol <aleixpol@kde.org>
Date: Thu, 28 Oct 2021 15:35:09 +0200
Subject: [PATCH] Add support for Primary Outputs

Introduce a kde_primary_output_v1 protocol to query
Changes kde_output_configuration_v2 to be able to change which is the
new
primary output.
---
 src/CMakeLists.txt                         |  1 +
 src/protocols/kde-output-device-v2.xml     |  9 ++++++++-
 src/protocols/kde-output-management-v2.xml |  8 ++++++--
 src/protocols/kde-primary-output-v1.xml    | 22 ++++++++++++++++++++++
 4 files changed, 37 insertions(+), 3 deletions(-)
 create mode 100644 src/protocols/kde-primary-output-v1.xml

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index c6433f92..9e5c96eb 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -9,6 +9,7 @@ install(FILES
     protocols/keystate.xml
     protocols/outputdevice.xml
     protocols/output-management.xml
+    protocols/kde-primary-output-v1.xml
     protocols/kde-output-device-v2.xml
     protocols/kde-output-management-v2.xml
     protocols/plasma-shell.xml
diff --git a/src/protocols/kde-output-device-v2.xml b/src/protocols/kde-output-device-v2.xml
index fb6b9b02..f2bb1a3d 100644
--- a/src/protocols/kde-output-device-v2.xml
+++ b/src/protocols/kde-output-device-v2.xml
@@ -11,7 +11,7 @@
     ]]></copyright>
 
 
-  <interface name="kde_output_device_v2" version="1">
+  <interface name="kde_output_device_v2" version="2">
     <description summary="output configuration representation">
       An output device describes a display device available to the compositor.
       output_device is similar to wl_output, but focuses on output
@@ -266,6 +266,13 @@
       </description>
       <arg name="rgb_range" type="uint" enum="rgb_range"/>
     </event>
+
+    <event name="name" since="2">
+      <description summary="Output's name">
+        Name of the output, it's useful to cross-reference to an zxdg_output_v1 and ultimately QScreen
+      </description>
+      <arg name="name" type="string"/>
+    </event>
   
   </interface>
 
diff --git a/src/protocols/kde-output-management-v2.xml b/src/protocols/kde-output-management-v2.xml
index 58f59f97..605747fb 100644
--- a/src/protocols/kde-output-management-v2.xml
+++ b/src/protocols/kde-output-management-v2.xml
@@ -10,7 +10,7 @@
     SPDX-License-Identifier: MIT-CMU
     ]]></copyright>
 
-<interface name="kde_output_management_v2" version="1">
+<interface name="kde_output_management_v2" version="2">
   <description summary="configuration of server outputs through clients">
     This interface enables clients to set properties of output devices for screen
     configuration purposes via the server. To this end output devices are referenced
@@ -62,7 +62,7 @@
 
 </interface>
 
-<interface name="kde_output_configuration_v2" version="1">
+<interface name="kde_output_configuration_v2" version="2">
   <description summary="configure single output devices">
     outputconfiguration is a client-specific resource that can be used to ask
     the server to apply changes to available output devices.
@@ -192,6 +192,10 @@
     <arg name="rgb_range" type="uint" enum="rgb_range"/>
   </request>
 
+  <request name="set_primary_output" since="2">
+    <description summary="Select which primary output to use" />
+    <arg name="output" type="object" interface="kde_output_device_v2" allow-null="false"/>
+  </request>
 </interface>
 
 </protocol>
diff --git a/src/protocols/kde-primary-output-v1.xml b/src/protocols/kde-primary-output-v1.xml
new file mode 100644
index 00000000..7da58eee
--- /dev/null
+++ b/src/protocols/kde-primary-output-v1.xml
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<protocol name="kde_primary_output_v1">
+  <copyright><![CDATA[
+    SPDX-FileCopyrightText: 2021 Aleix Pol Gonzalez <aleixpol@kde.org>
+
+    SPDX-License-Identifier: MIT-CMU
+    ]]></copyright>
+
+<interface name="kde_primary_output_v1" version="1">
+  <description summary="expose which is the primary display">
+      Protocol for telling which is the primary display among the selection of enabled outputs.
+  </description>
+
+  <event name="primary_output">
+    <description summary="Provide the current primary output's name">
+      Specifies which output is the primary one identified by their uuid. See kde_output_device_v2 uuid event for more information about it.
+    </description>
+    <arg name="output_name" type="string" summary="the name of the output"/>
+  </event>
+</interface>
+
+</protocol>
-- 
GitLab

